integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java

```diff
@@ -0,0 +1,484 @@
+            throw new Exception(connectorName + ": topic " + topicName + " not found");
+         }
+
+         this.retryCount.set(0);
+         this.isStarted.set(true);
+         connect();
+         ActiveMQKafkaLogger.LOGGER.bridgeStarted(connectorName);
+      }
+   }
+
+   public void connect() throws Exception {
+      synchronized (this) {
+         if (!isConnected && isStarted.get()) {
+            isConnected = true;
+            this.queue.addConsumer(this);
+            this.queue.deliverAsync();
+            ActiveMQKafkaLogger.LOGGER.bridgeConnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void disconnect() {
+      synchronized (this) {
+         if (isConnected) {
+            if (queue != null) {
+               this.queue.removeConsumer(this);
+            }
+
+            cancelRefs();
+            if (queue != null) {
+               queue.deliverAsync();
+            }
+            isConnected = false;
+            ActiveMQKafkaLogger.LOGGER.bridgeDisconnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void stop() {
+      synchronized (this) {
+         if (!this.isStarted.get()) {
+            return;
+         }
+         ActiveMQKafkaLogger.LOGGER.bridgeReceivedStopRequest(connectorName);
+
+         disconnect();
+
+         kafkaProducer.close();
+         kafkaProducer = null;
+         this.isStarted.set(false);
+         ActiveMQKafkaLogger.LOGGER.bridgeStopped(connectorName);
+      }
+   }
+
+   @Override
+   public boolean isStarted() {
+      return this.isStarted.get();
+   }
+
+   @Override
+   public String getName() {
+      return this.connectorName;
+   }
+
+   @Override
+   public boolean supportsDirectDelivery() {
+      return false;
+   }
+
+   @Override
+   public HandleStatus handle(MessageReference ref) throws Exception {
+      if (filter != null && !filter.match(ref.getMessage())) {
+         return HandleStatus.NO_MATCH;
+      }
+
+      synchronized (this) {
+         ref.handled();
+
+         Message message = ref.getMessage();
+
+         synchronized (refs) {
+            refs.put(ref.getMessage().getMessageID(), ref);
+         }
+
+         Integer partition = null;
+         SimpleString groupdID = message.getGroupID();
+         if (groupdID != null) {
+            List partitions = kafkaProducer.partitionsFor(topicName);
+            int numPartitions = partitions.size();
+            partition = Utils.toPositive(Utils.murmur2(groupdID.getData())) % numPartitions;
+         }
+
+         //Compaction/Record Key (equivalent to Artemis LVQ)
+         SimpleString key = message.getLastValueProperty();
+
+         ProducerRecord<String, Message> producerRecord =
+             new ProducerRecord<>(
+                   topicName,
+                   partition,
+                   message.getTimestamp(),
+                   key == null ? null : key.toString(),
+                   message);
```

Comment:
Providing both partition and key (if it's not null), the Kafka producer will always use partition which takes precedence on the use of the key. It means that if groupId is specified, the LVQ (as key) won't be consider for addressing the partition. Is it exactly what you had in mind ?

Reply:
Yes key, is used for compaction in kafka primarily, we use groupId if present to primarily partition by to keep similar semantics as consumers in kafka consume by partition there for a group is honoured.